File: packages/binding/android/src/main/java/com/pingidentity/rnbinding/BindingOptionParser.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package com.pingidentity.rnbinding

import androidx.annotation.VisibleForTesting
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.pingidentity.device.binding.Prompt
import com.pingidentity.device.binding.journey.DeviceBindingConfig
import com.pingidentity.rncore.CoreRuntime
import com.pingidentity.rncore.logger.LoggerHandleContract
import com.pingidentity.rncore.storage.StorageConfigHandleContract
import com.pingidentity.storage.CacheStrategy
import com.pingidentity.logger.Logger
import java.time.Instant

private const val LOGGER_ID_KEY = "loggerId"
private const val HAS_PIN_COLLECTOR_KEY = "hasPinCollector"
private const val HAS_USER_KEY_SELECTOR_KEY = "hasUserKeySelector"

/**
 * Parsed per-call runtime configuration extracted from the JS `config` argument.
 *
 * @property loggerId Optional registry ID of a logger registered via `RNPingCore`.
 * @property hasPinCollector True when the JS caller has registered a PIN collector callback.
 * @property hasUserKeySelector True when the JS caller has registered a user-key selector callback.
 * @property userKeyStorageId Optional registry ID of a user-key storage config registered via `RNPingCore`.
 */
internal data class CallConfig(
  val loggerId: String?,
  val hasPinCollector: Boolean = false,
  val hasUserKeySelector: Boolean = false,
  val userKeyStorageId: String? = null,
)

/**
 * Reads the `index` field from [options], supporting both numeric and string representations.
 * Defaults to `0` when the field is absent, null, or not parseable as an integer.
 */
internal fun parseCallbackIndex(options: ReadableMap?): Int {
  if (options == null || !options.hasKey("index") || options.isNull("index")) return 0
  return when (options.getType("index")) {
    ReadableType.Number -> options.getDouble("index").toInt()
    ReadableType.String -> options.getString("index")?.toIntOrNull() ?: 0
    else -> 0
  }
}

/**
 * Reads a trimmed, non-blank string value from [options] at [key].
 * Returns null when the key is absent, null, not a string, or blank after trimming.
 */
internal fun parseStringOption(options: ReadableMap?, key: String): String? {
  if (options == null || !options.hasKey(key) || options.isNull(key)) return null
  return when (options.getType(key)) {
    ReadableType.String -> options.getString(key)?.trim()?.takeIf { it.isNotEmpty() }
    else -> null
  }
}

/**
 * Extracts the `claims` map from [options] as a flat `Map<String, Any>`.
 * Only string, number, and boolean leaf values are included; nested maps and arrays are ignored.
 * Returns an empty map when the field is absent or not a map.
 */
internal fun parseClaims(options: ReadableMap?): Map<String, Any> {
  if (options == null || !options.hasKey("claims") || options.isNull("claims")) return emptyMap()
  if (options.getType("claims") != ReadableType.Map) return emptyMap()
  val claimsMap = options.getMap("claims") ?: return emptyMap()
  val result = mutableMapOf<String, Any>()
  val iterator = claimsMap.keySetIterator()
  while (iterator.hasNextKey()) {
    val key = iterator.nextKey()
    when (claimsMap.getType(key)) {
      ReadableType.String -> claimsMap.getString(key)?.let { result[key] = it }
      ReadableType.Number -> result[key] = claimsMap.getDouble(key)
      ReadableType.Boolean -> result[key] = claimsMap.getBoolean(key)
      else -> {}
    }
  }
  return result
}

/**
 * Parses the JS `config` argument into a [CallConfig].
 *
 * @param config The per-call configuration map passed from JS.
 * @return A [CallConfig] with all optional fields resolved.
 */
internal fun parseConfig(config: ReadableMap?): CallConfig {
  val loggerId = if (config != null && config.hasKey(LOGGER_ID_KEY) && !config.isNull(LOGGER_ID_KEY)) {
    when (config.getType(LOGGER_ID_KEY)) {
      ReadableType.String -> config.getString(LOGGER_ID_KEY)?.trim()?.takeIf { it.isNotEmpty() }
      else -> null
    }
  } else null

  val hasPinCollector = config != null &&
    config.hasKey(HAS_PIN_COLLECTOR_KEY) &&
    !config.isNull(HAS_PIN_COLLECTOR_KEY) &&
    config.getType(HAS_PIN_COLLECTOR_KEY) == ReadableType.Boolean &&
    config.getBoolean(HAS_PIN_COLLECTOR_KEY)

  val hasUserKeySelector = config != null &&
    config.hasKey(HAS_USER_KEY_SELECTOR_KEY) &&
    !config.isNull(HAS_USER_KEY_SELECTOR_KEY) &&
    config.getType(HAS_USER_KEY_SELECTOR_KEY) == ReadableType.Boolean &&
    config.getBoolean(HAS_USER_KEY_SELECTOR_KEY)

  return CallConfig(
    loggerId = loggerId,
    hasPinCollector = hasPinCollector,
    hasUserKeySelector = hasUserKeySelector,
    userKeyStorageId = parseStringOption(config, "userKeyStorageId"),
  )
}

/**
 * Resolves a [Logger] from the `CoreRuntime` registry using [loggerId].
 * Returns null when [loggerId] is blank or no matching handle is registered.
 */
internal fun resolveLoggerFromCore(loggerId: String?): Logger? {
  if (loggerId.isNullOrBlank()) return null
  val handle = CoreRuntime.loggerRegistry.resolve(loggerId) as? LoggerHandleContract ?: return null
  return handle.nativeLogger as? Logger
}

/**
 * Applies JS options (appPin, biometric, jwt) and optional user-key storage config to
 * [bindingConfig]. This mutates the config in-place inside a `bind {}` or `sign {}` block.
 *
 * @param bindingConfig The native [DeviceBindingConfig] to configure.
 * @param options The JS options map containing appPin, biometric, and jwt sub-keys.
 * @param userKeyStorageId Optional registry ID for a user-key storage config.
 */
internal fun applyCommonBindingConfig(
  bindingConfig: DeviceBindingConfig,
  options: ReadableMap?,
  userKeyStorageId: String?
) {
  userKeyStorageId?.let { id ->
    val handle = CoreRuntime.bindingUserKeyStorageConfigRegistry.resolve(id)
    val storageHandle = handle as? StorageConfigHandleContract
    if (storageHandle != null) {
      bindingConfig.userKeyStorage {
        storage {
          storageHandle.fileName?.let { fileName = it }
          storageHandle.keyAlias?.let { keyAlias = it }
          storageHandle.strongBoxPreferred?.let { strongBoxPreferred = it }
          storageHandle.cacheStrategy?.let { value -> cacheStrategy = parseCacheStrategy(value) }
        }
      }
    }
  }
  if (options == null) return

  val appPin = options.getMap("appPin")
  if (appPin != null) {
    bindingConfig.appPinConfig {
      if (appPin.hasKey("maxAttempts") && !appPin.isNull("maxAttempts")) {
        pinRetry = appPin.getDouble("maxAttempts").toInt()
      }
      parseStringOption(appPin, "keystoreType")?.let { keystoreType = it }
      val promptMap = appPin.getMap("prompt")
      if (promptMap != null) {
        val title = if (promptMap.hasKey("title") && !promptMap.isNull("title")) {
          promptMap.getString("title") ?: ""
        } else prompt.title
        val subtitle = if (promptMap.hasKey("subtitle") && !promptMap.isNull("subtitle")) {
          promptMap.getString("subtitle") ?: ""
        } else prompt.subtitle
        val description = if (promptMap.hasKey("description") && !promptMap.isNull("description")) {
          promptMap.getString("description") ?: ""
        } else prompt.description
        prompt = Prompt(title = title, subtitle = subtitle, description = description)
      }
    }
  }

  val androidBiometric = options.getMap("biometric")?.getMap("android")
  if (androidBiometric != null) {
    val promptMap = androidBiometric.getMap("prompt")
    val jsStrongBoxPreferred = if (androidBiometric.hasKey("strongBoxPreferred") && !androidBiometric.isNull("strongBoxPreferred")) {
      androidBiometric.getBoolean("strongBoxPreferred")
    } else null
    if (promptMap != null || jsStrongBoxPreferred != null) {
      bindingConfig.biometricAuthenticatorConfig {
        jsStrongBoxPreferred?.let { strongBoxPreferred = it }
        if (promptMap != null) {
          val promptTitle = parseStringOption(promptMap, "title")
          val promptSubtitle = parseStringOption(promptMap, "subtitle")
          val promptDescription = parseStringOption(promptMap, "description")
          val promptNegativeButton = parseStringOption(promptMap, "negativeButtonText")
          if (promptTitle != null || promptSubtitle != null || promptDescription != null || promptNegativeButton != null) {
            promptInfo {
              promptTitle?.let { setTitle(it) }
              promptSubtitle?.let { setSubtitle(it) }
              promptDescription?.let { setDescription(it) }
              promptNegativeButton?.let { setNegativeButtonText(it) }
            }
          }
        }
      }
    }
  }

  val jwt = options.getMap("jwt")
  if (jwt != null) {
    if (jwt.hasKey("issueTimeEpochSeconds") && !jwt.isNull("issueTimeEpochSeconds")) {
      val epoch = jwt.getDouble("issueTimeEpochSeconds").toLong()
      bindingConfig.issueTime = { Instant.ofEpochSecond(epoch) }
    }
    if (jwt.hasKey("notBeforeTimeEpochSeconds") && !jwt.isNull("notBeforeTimeEpochSeconds")) {
      val epoch = jwt.getDouble("notBeforeTimeEpochSeconds").toLong()
      bindingConfig.notBeforeTime = { Instant.ofEpochSecond(epoch) }
    }
    if (jwt.hasKey("expirationTimeEpochSeconds") && !jwt.isNull("expirationTimeEpochSeconds")) {
      val epoch = jwt.getDouble("expirationTimeEpochSeconds").toLong()
      bindingConfig.expirationTime = { Instant.ofEpochSecond(epoch) }
    }
  }
}

private fun parseCacheStrategy(value: String): CacheStrategy = when (value) {
  "cache" -> CacheStrategy.CACHE
  "cache_on_failure" -> CacheStrategy.CACHE_ON_FAILURE
  else -> CacheStrategy.NO_CACHE
}

// ─── Test support ────────────────────────────────────────────────────────────

@VisibleForTesting
internal data class ParsedBindingOptions(
  val appPinMaxAttempts: Int?,
  val appPinKeystoreType: String?,
  val appPinPromptTitle: String?,
  val appPinPromptSubtitle: String?,
  val appPinPromptDescription: String?,
  val biometricAndroidStrongBoxPreferred: Boolean?,
  val biometricAndroidPromptTitle: String?,
  val biometricAndroidPromptSubtitle: String?,
  val biometricAndroidPromptDescription: String?,
  val biometricAndroidPromptNegativeButtonText: String?,
  val jwtIssueTimeEpochSeconds: Long?,
  val jwtNotBeforeTimeEpochSeconds: Long?,
  val jwtExpirationTimeEpochSeconds: Long?,
)

@VisibleForTesting
internal fun parseBindingOptionsForTest(options: ReadableMap?): ParsedBindingOptions {
  if (options == null) return ParsedBindingOptions(
    null, null, null, null, null, null, null, null, null, null, null, null, null
  )
  val appPin = options.getMap("appPin")
  val appPinMaxAttempts = if (appPin != null && appPin.hasKey("maxAttempts") && !appPin.isNull("maxAttempts")) {
    appPin.getDouble("maxAttempts").toInt()
  } else null
  val promptMap = appPin?.getMap("prompt")

  val androidBiometric = options.getMap("biometric")?.getMap("android")
  val biometricAndroidStrongBoxPreferred = if (
    androidBiometric != null &&
    androidBiometric.hasKey("strongBoxPreferred") &&
    !androidBiometric.isNull("strongBoxPreferred")
  ) androidBiometric.getBoolean("strongBoxPreferred") else null
  val biometricPromptMap = androidBiometric?.getMap("prompt")

  val jwt = options.getMap("jwt")
  return ParsedBindingOptions(
    appPinMaxAttempts = appPinMaxAttempts,
    appPinKeystoreType = parseStringOption(appPin, "keystoreType"),
    appPinPromptTitle = parseStringOption(promptMap, "title"),
    appPinPromptSubtitle = parseStringOption(promptMap, "subtitle"),
    appPinPromptDescription = parseStringOption(promptMap, "description"),
    biometricAndroidStrongBoxPreferred = biometricAndroidStrongBoxPreferred,
    biometricAndroidPromptTitle = parseStringOption(biometricPromptMap, "title"),
    biometricAndroidPromptSubtitle = parseStringOption(biometricPromptMap, "subtitle"),
    biometricAndroidPromptDescription = parseStringOption(biometricPromptMap, "description"),
    biometricAndroidPromptNegativeButtonText = parseStringOption(biometricPromptMap, "negativeButtonText"),
    jwtIssueTimeEpochSeconds = if (jwt != null && jwt.hasKey("issueTimeEpochSeconds") && !jwt.isNull("issueTimeEpochSeconds")) jwt.getDouble("issueTimeEpochSeconds").toLong() else null,
    jwtNotBeforeTimeEpochSeconds = if (jwt != null && jwt.hasKey("notBeforeTimeEpochSeconds") && !jwt.isNull("notBeforeTimeEpochSeconds")) jwt.getDouble("notBeforeTimeEpochSeconds").toLong() else null,
    jwtExpirationTimeEpochSeconds = if (jwt != null && jwt.hasKey("expirationTimeEpochSeconds") && !jwt.isNull("expirationTimeEpochSeconds")) jwt.getDouble("expirationTimeEpochSeconds").toLong() else null,
  )
}

File: packages/binding/ios/BindingOptionParser.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
import Foundation
import PingBinding
import PingStorage
import PingLogger
import RNPingCore

/// Parsed app-pin authenticator options from the JS `options.appPin` sub-key.
struct ParsedAppPinOptions: Sendable {
  let maxAttempts: Int?
  let keyTag: String?
  let promptTitle: String?
  let promptSubtitle: String?
  let promptDescription: String?
}

/// Parsed biometric authenticator options from the JS `options.biometric.ios` sub-key.
struct ParsedBiometricOptions: Sendable {
  let keyTag: String?
}

/// Parsed JWT claim timing options from the JS `options.jwt` sub-key.
struct ParsedJwtOptions: Sendable {
  let issueTimeEpochSeconds: Double?
  let notBeforeTimeEpochSeconds: Double?
  let expirationTimeEpochSeconds: Double?
}

/// Resolved user-key storage configuration fetched from the `CoreRuntime` registry.
struct ResolvedUserKeyStorageValues: Sendable {
  /// The Keychain account string for user-key persistence.
  let account: String
  /// Whether to wrap the Keychain entry with `SecuredKeyEncryptor`.
  let encryptorEnabled: Bool
}

extension RNPingBindingCommon {

  /// Reads the `index` field from `options`, accepting both `NSNumber` and `String` representations.
  /// Returns `0` when the field is absent or not parseable.
  static func parseCallbackIndex(_ options: NSDictionary) -> Int {
    if let value = options["index"] as? NSNumber { return value.intValue }
    if let value = options["index"] as? String, let parsed = Int(value) { return parsed }
    return 0
  }

  /// Reads a trimmed, non-blank string from `options` at `key`. Returns `nil` when absent or blank.
  static func parseStringOption(_ options: NSDictionary, key: String) -> String? {
    guard let value = options[key] as? String else { return nil }
    let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return normalized.isEmpty ? nil : normalized
  }

  /// Reads a trimmed, non-blank string from the per-call `config` dict at `key`. Returns `nil` when absent or blank.
  static func parseStringConfig(_ config: NSDictionary, key: String) -> String? {
    guard let value = config[key] as? String else { return nil }
    let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return normalized.isEmpty ? nil : normalized
  }

  /// Reads the `loggerId` field from the per-call `config` dict. Returns `nil` when absent or blank.
  static func parseLoggerId(_ config: NSDictionary) -> String? {
    guard let value = config["loggerId"] as? String else { return nil }
    let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return normalized.isEmpty ? nil : normalized
  }

  /// Extracts the `claims` dictionary from `options`. Returns an empty dict when absent or wrong type.
  static func parseClaims(_ options: NSDictionary) -> [String: Any] {
    guard let claims = options["claims"] as? [String: Any] else { return [:] }
    return claims
  }

  /// Parses `options.appPin` into `ParsedAppPinOptions`. Returns `nil` when the key is absent.
  static func parseAppPinOptions(_ options: NSDictionary) -> ParsedAppPinOptions? {
    guard let appPin = options["appPin"] as? NSDictionary else { return nil }
    let prompt = appPin["prompt"] as? NSDictionary
    return ParsedAppPinOptions(
      maxAttempts: (appPin["maxAttempts"] as? NSNumber)?.intValue,
      keyTag: appPin["keyTag"] as? String,
      promptTitle: prompt?["title"] as? String,
      promptSubtitle: prompt?["subtitle"] as? String,
      promptDescription: prompt?["description"] as? String
    )
  }

  /// Parses `options.biometric.ios` into `ParsedBiometricOptions`. Returns `nil` when the key is absent.
  static func parseBiometricOptions(_ options: NSDictionary) -> ParsedBiometricOptions? {
    guard let biometric = options["biometric"] as? NSDictionary,
          let ios = biometric["ios"] as? NSDictionary else { return nil }
    return ParsedBiometricOptions(
      keyTag: ios["keyTag"] as? String
    )
  }

  /// Parses `options.jwt` into `ParsedJwtOptions`. Returns `nil` when the key is absent.
  static func parseJwtOptions(_ options: NSDictionary) -> ParsedJwtOptions? {
    guard let jwt = options["jwt"] as? NSDictionary else { return nil }
    return ParsedJwtOptions(
      issueTimeEpochSeconds: (jwt["issueTimeEpochSeconds"] as? NSNumber)?.doubleValue,
      notBeforeTimeEpochSeconds: (jwt["notBeforeTimeEpochSeconds"] as? NSNumber)?.doubleValue,
      expirationTimeEpochSeconds: (jwt["expirationTimeEpochSeconds"] as? NSNumber)?.doubleValue
    )
  }

  /// Resolves a `PingLogger.Logger` from the `CoreRuntime` registry using `loggerId`.
  /// Returns `nil` when `loggerId` is blank or no matching handle is registered.
  static func resolveLoggerFromCore(_ loggerId: String?) async -> PingLogger.Logger? {
    guard let loggerId, !loggerId.isEmpty else { return nil }
    guard let handle = await CoreRuntime.loggerRegistry.resolve(loggerId) as? LoggerHandleContract else { return nil }
    return handle.nativeLogger as? PingLogger.Logger
  }

  /// Sets the `logger` property on `config` via KVC when the config type supports it.
  /// Falls back to `LogManager.none` when `logger` is `nil`.
  static func setLoggerIfSupported(_ config: AnyObject, logger: PingLogger.Logger?) {
    guard let target = config as? NSObject else { return }
    let selector = NSSelectorFromString("setLogger:")
    guard target.responds(to: selector) else { return }
    target.setValue(logger ?? LogManager.none, forKey: "logger")
  }

  /// Resolves user-key storage configuration from the `CoreRuntime` registry using `userKeyStorageId`.
  /// Returns `nil` when `userKeyStorageId` is blank or no matching handle is registered.
  static func resolveUserKeyStorageFromCore(_ userKeyStorageId: String?) async -> ResolvedUserKeyStorageValues? {
    guard let userKeyStorageId, !userKeyStorageId.isEmpty else { return nil }
    guard let handle = await CoreRuntime.bindingUserKeyStorageConfigRegistry.resolve(userKeyStorageId)
      as? StorageConfigHandleContract else { return nil }
    let rawAccount = handle.account?.trimmingCharacters(in: .whitespacesAndNewlines)
    let resolvedAccount = (rawAccount?.isEmpty == false)
      ? (rawAccount ?? "com.pingidentity.device.binding.v1.userkeys")
      : "com.pingidentity.device.binding.v1.userkeys"
    return ResolvedUserKeyStorageValues(
      account: resolvedAccount,
      encryptorEnabled: handle.encryptor ?? true
    )
  }

  /// Applies parsed options to `nativeConfig` inside a `bind {}` or `sign {}` closure.
  ///
  /// - Parameters:
  ///   - nativeConfig: The `DeviceBindingConfig` to mutate.
  ///   - appPin: Parsed app-pin options, or `nil` to skip.
  ///   - biometric: Parsed biometric options, or `nil` to skip.
  ///   - jwt: Parsed JWT timing options, or `nil` to skip.
  ///   - resolvedUserKeyStorage: Resolved user-key storage values, or `nil` to skip.
  static func applyCommonBindingConfig(
    nativeConfig: DeviceBindingConfig,
    appPin: ParsedAppPinOptions?,
    biometric: ParsedBiometricOptions?,
    jwt: ParsedJwtOptions?,
    resolvedUserKeyStorage: ResolvedUserKeyStorageValues?
  ) {
    if let storage = resolvedUserKeyStorage {
      let encryptor: any Encryptor
      if storage.encryptorEnabled {
        encryptor = SecuredKeyEncryptor() ?? NoEncryptor()
      } else {
        encryptor = NoEncryptor()
      }
      nativeConfig.userKeyStorage = UserKeyStorageConfig(
        storage: KeychainStorage(account: storage.account, encryptor: encryptor)
      )
    }
    if let appPin { applyAppPinConfig(nativeConfig: nativeConfig, appPin: appPin) }
    if let biometric { applyBiometricConfig(nativeConfig: nativeConfig, biometric: biometric) }
    if let jwt { applyJwtConfig(nativeConfig: nativeConfig, jwt: jwt) }
  }

  static func configureAppPinConfig(_ config: AppPinConfig, appPin: ParsedAppPinOptions) {
    if let maxAttempts = appPin.maxAttempts { config.pinRetry = maxAttempts }
    if let keyTag = appPin.keyTag, !keyTag.isEmpty { config.keyTag = keyTag }
    if appPin.promptTitle != nil || appPin.promptSubtitle != nil || appPin.promptDescription != nil {
      config.prompt = Prompt(
        title: appPin.promptTitle ?? config.prompt.title,
        subtitle: appPin.promptSubtitle ?? config.prompt.subtitle,
        description: appPin.promptDescription ?? config.prompt.description
      )
    }
  }

  static func applyAppPinConfig(nativeConfig: DeviceBindingConfig, appPin: ParsedAppPinOptions) {
    let defaultConfig = AppPinConfig()
    configureAppPinConfig(defaultConfig, appPin: appPin)
    nativeConfig.authenticatorConfig = defaultConfig
  }

  static func applyBiometricConfig(nativeConfig: DeviceBindingConfig, biometric: ParsedBiometricOptions) {
    let biometricConfig = BiometricAuthenticatorConfig()
    if let keyTag = biometric.keyTag, !keyTag.isEmpty { biometricConfig.keyTag = keyTag }
    nativeConfig.authenticatorConfig = biometricConfig
  }

  static func applyJwtConfig(nativeConfig: DeviceBindingConfig, jwt: ParsedJwtOptions) {
    if let issue = jwt.issueTimeEpochSeconds {
      let date = Date(timeIntervalSince1970: issue)
      nativeConfig.issueTime = { date }
    }
    if let notBefore = jwt.notBeforeTimeEpochSeconds {
      let date = Date(timeIntervalSince1970: notBefore)
      nativeConfig.notBeforeTime = { date }
    }
    if let expiration = jwt.expirationTimeEpochSeconds {
      let date = Date(timeIntervalSince1970: expiration)
      nativeConfig.expirationTime = { _ in date }
    }
  }

#if DEBUG
  public static func _test_configureAppPinConfig(_ options: NSDictionary) -> NSDictionary? {
    guard let appPin = parseAppPinOptions(options) else { return nil }
    let config = AppPinConfig()
    configureAppPinConfig(config, appPin: appPin)
    return [
      "pinRetry": config.pinRetry as Any,
      "keyTag": config.keyTag as Any,
      "promptTitle": config.prompt.title as Any,
      "promptSubtitle": config.prompt.subtitle as Any,
      "promptDescription": config.prompt.description as Any,
    ]
  }

  public static func _test_parseAppPinOptions(_ options: NSDictionary) -> NSDictionary? {
    guard let parsed = parseAppPinOptions(options) else { return nil }
    return [
      "maxAttempts": parsed.maxAttempts as Any,
      "keyTag": parsed.keyTag as Any,
      "promptTitle": parsed.promptTitle as Any,
      "promptSubtitle": parsed.promptSubtitle as Any,
      "promptDescription": parsed.promptDescription as Any,
    ]
  }

  public static func _test_parseBiometricOptions(_ options: NSDictionary) -> NSDictionary? {
    guard let parsed = parseBiometricOptions(options) else { return nil }
    return [
      "keyTag": parsed.keyTag as Any,
    ]
  }

  public static func _test_parseJwtOptions(_ options: NSDictionary) -> NSDictionary? {
    guard let parsed = parseJwtOptions(options) else { return nil }
    return [
      "issueTimeEpochSeconds": parsed.issueTimeEpochSeconds as Any,
      "notBeforeTimeEpochSeconds": parsed.notBeforeTimeEpochSeconds as Any,
      "expirationTimeEpochSeconds": parsed.expirationTimeEpochSeconds as Any,
    ]
  }

  public static func _test_parseClaims(_ options: NSDictionary) -> NSDictionary {
    return parseClaims(options) as NSDictionary
  }

  public static func _test_parseCallbackIndex(_ options: NSDictionary) -> Int {
    return parseCallbackIndex(options)
  }

  public static func _test_parseStringOption(_ options: NSDictionary, key: String) -> String? {
    return parseStringOption(options, key: key)
  }

  public static func _test_parseLoggerId(_ config: NSDictionary) -> String? {
    return parseLoggerId(config)
  }

  public static func _test_parseStringConfig(_ config: NSDictionary, key: String) -> String? {
    return parseStringConfig(config, key: key)
  }
#endif
}
